haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -229,45 +303,382 @@ def run(  # noqa: PLR0915, PLR0912
+        for bp in matching_breakpoints:
+            visit_count = bp[1]
+            # break at every visit if visit_count is -1
+            if visit_count == -1:
+                msg = f"Breaking at component: {component_name}"
+                logger.info(msg)
+                state = self.save_state(inputs, str(component_name), instance, component_visits)
+                raise PipelineBreakException(msg, component=component_name, state=state)
+
+            # break only if the visit count is the same
+            if bp[1] == component_visits[component_name]:
+                msg = f"Breaking at component {component_name} visit count {component_visits[component_name]}"
+                logger.info(msg)
+                state = self.save_state(inputs, str(component_name), instance, component_visits)
+                raise PipelineBreakException(msg, component=component_name, state=state)
+
+    @staticmethod
+    def _remove_unserialisable_data(value):
+        """
+        Removes certain unserialisable data which is not needed for the pipeline state.
+        """
+
+        if isinstance(value, ChatMessage):  # noqa: SIM102
+            if "usage" in value.meta:  # noqa: SIM102
+                value.meta["usage"].pop("completion_tokens_details", None)
+                value.meta["usage"].pop("prompt_tokens_details", None)
+
+        if isinstance(value, GeneratedAnswer):  # noqa: SIM102
+            if value.meta and "usage" in value.meta:  # noqa: SIM102
+                value.meta.pop("usage", None)
+
+        return value
+
+    @staticmethod
+    def transform_json_structure(data, component):
+        """
+        Transforms a JSON structure by removing the 'sender' key and moving the 'value' to the top level.
+
+        For example:
+        "key": [{"sender": null, "value": "some value"}] -> "key": "some value"
+        """
+        if isinstance(data, dict):
+            # If this dict has both 'sender' and 'value', return just the value
+            if "value" in data and "sender" in data:
+                return data["value"]
+            # Otherwise, recursively process each key-value pair
+            return {k: Pipeline.transform_json_structure(v, component) for k, v in data.items()}
+
+        elif isinstance(data, list):
+            # First, transform each item in the list.
+            transformed = [Pipeline.transform_json_structure(item, component) for item in data]
+            # If the original list has exactly one element and that element was a dict
+            # with 'sender' and 'value', then unwrap the list.
+            if len(data) == 1 and isinstance(data[0], dict) and "value" in data[0] and "sender" in data[0]:
+                return transformed[0]
+            return transformed
+
+        else:
+            # For other data types, just return the value as is.
+            return data
+
+    @staticmethod
+    def _serialize_component_input(value, component):
```

Comment:
Let's use type annotations for the input params and returns here too. They make it easier to understand the code.